docs/architecture/dapr-for-net-developers/serviceinvocation.md

```diff
@@ -0,0 +1,356 @@
+
+![Update basket sequence diagram](./media/service-invocation/updatebasket.png)
+
+**Figure 6-4**. Update shopping basket sequence.
+
+Note how the aggregator service first retrieves catalog items from the Catalog API. It then validates items availability and pricing. Finally, the aggregator service saves the updated shopping basket by calling the Basket API.
+
+Let's look at some code. The aggregator service contains a `BasketController` that provides an endpoint for updating the shopping basket:
+
+``` csharp
+[Route("api/v1/[controller]")]
+[Authorize]
+[ApiController]
+public class BasketController : ControllerBase
+{
+    private readonly ICatalogService _catalog;
+    private readonly IBasketService _basket;
+
+    ...
+
+    [HttpPost]
+    [HttpPut]
+    public async Task<ActionResult<BasketData>> UpdateAllBasketAsync(
+        [FromBody] UpdateBasketRequest data, [FromHeader] string authorization)
+    {
+        // Get the item details from the catalog API.
+        var catalogItems = await _catalog.GetCatalogItemsAsync(
+            data.Items.Select(x => x.ProductId));
+
+        // Check item availability and prices; store results in basket object.
+        var basket = CreateValidatedBasket(data, catalogItems);
+
+        // Save the shopping basket.
+        await _basket.UpdateAsync(basket, authorization);
+
+        return basket;
+    }
+
+    ...
+```
+
+The `UpdateAllBasketAsync` method gets the *Authorization* header of the incoming request using a `FromHeader` attribute. The *Authorization* header contains the access token that is needed to call protected back-end services.
+
+After receiving a request to update the basket, the aggregator service calls the Catalog API to get the item details. The Basket controller uses an injected `ICatalogService` object to make that call and communicate with the Catalog API. The original implementation of the interface used gRPC to make the call. We changed the implementation to use Dapr service invocation:
+
+``` csharp
+public class CatalogService : ICatalogService
+{
+    private const string DaprAppId = "catalog-api";
+
+    private readonly DaprClient _daprClient;
+
+    public CatalogService(DaprClient daprClient)
+    {
+        _daprClient = daprClient;
+    }
+
+    public async Task<IEnumerable<CatalogItem>> GetCatalogItemsAsync(
+        IEnumerable<int> ids)
+    {
+        return await _daprClient.InvokeMethodAsync<IEnumerable<CatalogItem>>(
+            DaprAppId,
+            "api/v1/catalog/items",
+            new HTTPExtension
+            {
+                QueryString = new Dictionary<string, string>
+                {
+                    ["ids"] = string.Join(",", ids),
+                },
+                Verb = HTTPVerb.Get
+            });
+    }
```

Comment:
Async/await is necessary here because `_daprClient.InvokeMethodAsync` returns `ValueTask` whereas the `GetCatalogItemsAsync` must return `Task` (don't want to break pre-existing interfaces in eShop). An alternative would be to use the `AsTask()` method of `ValueTask`, which I also think is a bit confusing for less experienced developers. Anyway, I've opened an issue on the Dapr .NET SDK repo to replace `ValueTask` with `Task` because I don't think the use of `ValueTask` is beneficial in this case. If that proposal is accepted, I can remove the async/await from this code sample.